ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -482,68 +492,89 @@ KEXT_STATIC int HandleFileOpOperation(
         
         bool isDirectory = (0 != vnode_isdir(currentVnode));
         
-        VirtualizationRootHandle root = RootHandle_None;
+        VirtualizationRootHandle toRoot = RootHandle_None, fromRoot = RootHandle_None;
         FsidInode vnodeFsidInode = {};
-        int pid;
-        if (!ShouldHandleFileOpEvent(
+        int pid = -1;
+        bool messageDestinationProvider = ShouldHandleFileOpEvent(
                 &perfTracer,
                 context,
                 currentVnode,
                 action,
                 isDirectory,
-                &root,
+                fromPath,
+                &toRoot,
+                &fromRoot,
                 &vnodeFsidInode,
-                &pid))
+                &pid);
+        bool messageSourceProvider = VirtualizationRoot_IsValidRootHandle(fromRoot);
+        
+        if (!messageDestinationProvider && !messageSourceProvider)
         {
             goto CleanupAndReturn;
         }
         
         char procname[MAXCOMLEN + 1];
         proc_name(pid, procname, MAXCOMLEN + 1);
 
+        MessageType messageType;
+        PrjFSPerfCounter perfCounter;
         if (KAUTH_FILEOP_RENAME == action)
         {
-            PerfSample renameSample(&perfTracer, PrjFSPerfCounter_FileOp_Renamed);
+            perfCounter = PrjFSPerfCounter_FileOp_Renamed;
             
-            MessageType messageType =
+            messageType =
                 isDirectory
                 ? MessageType_KtoU_NotifyDirectoryRenamed
                 : MessageType_KtoU_NotifyFileRenamed;
-
+        }
+        else
+        {
+            perfCounter = PrjFSPerfCounter_FileOp_HardLinkCreated;
+            
+            messageType = MessageType_KtoU_NotifyFileHardLinkCreated;
+        }
+        
+        if (messageDestinationProvider)
```

Comment:
It looks like we don't actually send messages if !messageDestinationProvider.  Is there a chance a message will be sent later on for this case?

Reply:
I don't understand the question, can you please be more specific about the case you're thinking about? The code sends a message to the source provider if `messageSourceProvider` is true a few lines further down, i.e. we send a message to either the source or the destination provider. We don't want to message the destination provider if it is e.g. offline or the destination is outside a root.

Whether this is the behaviour we want is another interesting question of course, but as far as I can tell this is the original inspiration for the change, so I think the answer is yes.

Reply:
Nevermind, pulling the branch it all makes sense now.